docker/docker_image_dest.go

```diff
@@ -701,34 +702,49 @@ func (d *dockerImageDestination) putSignaturesToSigstoreAttachments(ctx context.
 		return errors.New("writing sigstore attachments is disabled by configuration")
 	}
 
-	ociManifest, err := d.c.getSigstoreAttachmentManifest(ctx, d.ref, manifestDigest)
+	genManifest, err := d.c.getSigstoreAttachmentManifest(ctx, d.ref, manifestDigest)
 	if err != nil {
 		return err
 	}
 	var ociConfig imgspecv1.Image // Most fields empty by default
-	if ociManifest == nil {
-		ociManifest = manifest.OCI1FromComponents(imgspecv1.Descriptor{
+	if genManifest == nil {
+		genManifest = manifest.OCI1FromComponents(imgspecv1.Descriptor{
 			MediaType: imgspecv1.MediaTypeImageConfig,
 			Digest:    "", // We will fill this in later.
 			Size:      0,
 		}, nil)
 		ociConfig.RootFS.Type = "layers"
 	} else {
-		logrus.Debugf("Fetching sigstore attachment config %s", ociManifest.Config.Digest.String())
+		blobInfo := genManifest.ConfigInfo()
+		logrus.Debugf("Fetching sigstore attachment config %s", blobInfo.Digest.String())
+		descriptor := imgspecv1.Descriptor{
+			MediaType:   blobInfo.MediaType,
+			Digest:      blobInfo.Digest,
+			Size:        blobInfo.Size,
+			Annotations: blobInfo.Annotations,
+		}
 		// We don’t benefit from a real BlobInfoCache here because we never try to reuse/mount configs.
-		configBlob, err := d.c.getOCIDescriptorContents(ctx, d.ref, ociManifest.Config, iolimits.MaxConfigBodySize,
+		configBlob, err := d.c.getOCIDescriptorContents(ctx, d.ref, descriptor, iolimits.MaxConfigBodySize,
 			none.NoCache)
 		if err != nil {
 			return err
 		}
 		if err := json.Unmarshal(configBlob, &ociConfig); err != nil {
-			return fmt.Errorf("parsing sigstore attachment config %s in %s: %w", ociManifest.Config.Digest.String(),
+			return fmt.Errorf("parsing sigstore attachment config %s in %s: %w", blobInfo.Digest.String(),
 				d.ref.ref.Name(), err)
 		}
 	}
 
-	// To make sure we can safely append to the slices of ociManifest, without adding a remote dependency on the code that creates it.
-	ociManifest.Layers = slices.Clone(ociManifest.Layers)
+	// To make sure we can safely append to the slices of Manifest, without adding a remote dependency on the code that creates it.
+	switch v := genManifest.(type) {
```

Comment:
No; hard-coding implementation types is a an invisible dependency, and a trap for long-term maintenance.

At the very least, this package can build a private interface. More likely, we would need an (internal-only, because we can’t add methods to a public interface without breaking API) extension of the `manifest.Manifest` interface which adds an “add a layer” operation. (Compare `internal/manifest.List{,Public}`).